File: turboci/id/from_string_test.go
// Copyright 2025 The LUCI Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//	http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package id

import (
	"fmt"
	"testing"

	"go.chromium.org/luci/common/testing/truth/assert"
)

func ExampleFromString() {
	// A simple workplan ID
	wp, err := FromString("Lmy-workplan")
	if err != nil {
		panic(err)
	}
	fmt.Println(wp.GetWorkPlan().GetId())

	// A check on that workplan
	chk, err := FromString("Lmy-workplan:Cmy-check")
	if err != nil {
		panic(err)
	}
	fmt.Println(chk.GetCheck().GetId())

	// A stage on that workplan
	st, err := FromString("Lmy-workplan:Smy-stage")
	if err != nil {
		panic(err)
	}
	fmt.Println(st.GetStage().GetId())

	// An invalid string
	_, err = FromString("Lmy-workplan:Xwhatever")
	fmt.Println(err)

	// Output:
	// my-workplan
	// my-check
	// my-stage
	// token 1 in "Lmy-workplan:Xwhatever": unexpected key 'X'
}

func TestFromStringErrorConditions(t *testing.T) {
	t.Parallel()

	tests := []struct {
		name        string
		input       string
		expectedErr string
	}{
		{
			name:        "Empty input string",
			input:       "",
			expectedErr: "",
		},
		{
			name:        "Empty token in middle",
			input:       "Lwp::Cchk",
			expectedErr: `token 1 in "Lwp::Cchk" was empty?`,
		},
		{
			name:        "Empty token at end",
			input:       "Lwp:Cchk:",
			expectedErr: `token 2 in "Lwp:Cchk:" was empty?`,
		},
		{
			name:        "Invalid idx - non-integer",
			input:       "Lwp:Cchk:Rabc",
			expectedErr: `token 2 in "Lwp:Cchk:Rabc": bad idx: strconv.ParseInt: parsing "abc": invalid syntax`,
		},
		{
			name:        "Invalid idx - negative",
			input:       "Lwp:Cchk:R-1",
			expectedErr: `token 2 in "Lwp:Cchk:R-1": bad idx: must be > 0`,
		},
		{
			name:        "Invalid timestamp - malformed",
			input:       "Lwp:Cchk:Vbadts",
			expectedErr: `token 2 in "Lwp:Cchk:Vbadts": bad timestamp: needed seconds/nanos`,
		},
		{
			name:        "Invalid timestamp - non-integer seconds",
			input:       "Lwp:Cchk:Vabc/123",
			expectedErr: `token 2 in "Lwp:Cchk:Vabc/123": bad timestamp: parsing seconds "abc"`,
		},
		{
			name:        "Invalid timestamp - non-integer nanos",
			input:       "Lwp:Cchk:V123/abc",
			expectedErr: `token 2 in "Lwp:Cchk:V123/abc": bad timestamp: parsing nanos "abc"`,
		},
		{
			name:        "Missing required previous ID part - R without C",
			input:       "Lwp:R1",
			expectedErr: `token 1 in "Lwp:R1": expected "Check", got "WorkPlan"`,
		},
		{
			name:        "Unexpected key for current ID type - L after L",
			input:       "Lwp1:Lwp2",
			expectedErr: `token 1 in "Lwp1:Lwp2": unexpected key 'L'`,
		},
		{
			name:        "Unexpected key for current ID type - C after C",
			input:       "Lwp:Cchk1:Cchk2",
			expectedErr: `token 2 in "Lwp:Cchk1:Cchk2": expected "WorkPlan", got "Check"`,
		},
		{
			name:        "WorkPlan with only key, no trimmed part - L",
			input:       "L",
			expectedErr: "",
		},
		{
			name:        "Token with only key, no trimmed part - C",
			input:       "Lwp:C",
			expectedErr: `token 1 in "Lwp:C": unexpected empty token`,
		},
	}

	for _, test := range tests {
		t.Run(test.name, func(t *testing.T) {
			t.Parallel()
			_, err := FromString(test.input)
			if test.expectedErr == "" {
				assert.NoErr(t, err)
			} else {
				assert.ErrIsLike(t, err, test.expectedErr)
			}
		})
	}
}

File: milo/ui/src/chronicle/utils/id/from_string.test.ts
// Copyright 2026 The LUCI Authors.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { fromString } from './from_string';

describe('chronicle/utils/id/from_string', () => {
  it('parses simple WorkPlan', () => {
    const parsed = fromString('Lmy-workplan');
    expect(parsed.workPlan).toEqual({ id: 'my-workplan' });
  });

  it('parses Check on WorkPlan', () => {
    const parsed = fromString('Lmy-workplan:Cmy-check');
    expect(parsed.check).toEqual({
      id: 'my-check',
      workPlan: { id: 'my-workplan' },
    });
  });

  it('parses Stage on WorkPlan', () => {
    const parsed = fromString('Lmy-workplan:Smy-stage');
    expect(parsed.stage).toEqual({
      id: 'my-stage',
      workPlan: { id: 'my-workplan' },
      isWorknode: false,
    });
  });

  it('parses Worknode Stage on WorkPlan', () => {
    const parsed = fromString('Lmy-workplan:Nmy-stage');
    expect(parsed.stage).toEqual({
      id: 'my-stage',
      workPlan: { id: 'my-workplan' },
      isWorknode: true,
    });
  });

  it('parses StageAttempt correctly', () => {
    const parsed = fromString('Lwp:Sst:A123');
    expect(parsed.stageAttempt).toEqual({
      stage: {
        id: 'st',
        workPlan: { id: 'wp' },
        isWorknode: false,
      },
      idx: 123,
    });
  });

  it('returns empty object for empty string', () => {
    expect(fromString('')).toEqual({});
  });

  describe('error conditions', () => {
    it('rejects empty tokens in the middle', () => {
      expect(() => fromString('Lwp::Cchk')).toThrow(
        'token 1 in "Lwp::Cchk" was empty?',
      );
    });

    it('rejects empty tokens at the end', () => {
      expect(() => fromString('Lwp:Cchk:')).toThrow(
        'token 2 in "Lwp:Cchk:" was empty?',
      );
    });

    it('rejects invalid non-integer index', () => {
      expect(() => fromString('Lwp:Cchk:Rabc')).toThrow(
        'expected a positive integer',
      );
      expect(() => fromString('Lwp:Cchk:R123a')).toThrow(
        'expected a positive integer',
      );
    });

    it('rejects negative index', () => {
      expect(() => fromString('Lwp:Cchk:R-1')).toThrow('bad idx: must be > 0');
    });

    it('rejects zero index', () => {
      expect(() => fromString('Lwp:Cchk:R0')).toThrow('bad idx: must be > 0');
    });

    it('rejects malformed timestamps', () => {
      expect(() => fromString('Lwp:Cchk:Vbadts')).toThrow(
        'bad timestamp: needed seconds/nanos',
      );
    });

    it('rejects invalid seconds inside timestamp', () => {
      expect(() => fromString('Lwp:Cchk:Vabc/123')).toThrow(
        'bad timestamp: parsing seconds "abc"',
      );
      expect(() => fromString('Lwp:Cchk:V123a/123')).toThrow(
        'bad timestamp: parsing seconds "123a"',
      );
    });

    it('rejects invalid nanos inside timestamp', () => {
      expect(() => fromString('Lwp:Cchk:V123/abc')).toThrow(
        'bad timestamp: parsing nanos "abc"',
      );
    });

    it('rejects unexpected keys', () => {
      expect(() => fromString('Lwp1:Lwp2')).toThrow("unexpected key 'L'");
    });
  });
});
